substratevm/src/com.oracle.graal.pointsto/src/com/oracle/graal/pointsto/typestate/MultiTypeState.java

```diff
@@ -185,11 +266,78 @@ public boolean equals(Object o) {
 
         MultiTypeState that = (MultiTypeState) o;
         return this.canBeNull == that.canBeNull &&
-                        this.typesCount == that.typesCount && this.typesBitSet.equals(that.typesBitSet);
+                        this.typesCount == that.typesCount && this.bitSetEquals(that);
     }
 
     @Override
     public String toString() {
         return "MType<" + typesCount + ":" + (canBeNull ? "null," : "") + "TODO" + ">";
     }
+
+    class SmallBitSet {
+        public SmallBitSet(BitSet other) {
+            assert other.cardinality() <= MAX_CARDINALITY;
+            for (int i = other.nextSetBit(0); i >= 0; i = other.nextSetBit(i + 1)) {
+                set[cardinality++] = i;
+            }
+            assert cardinality == other.cardinality();
+        }
+
+        public boolean get(int bit) {
+            assert bit >= 0 : "get() bit must be non-negative";
+            for (int i = 0; i < cardinality; i++) {
+                if (set[i] == bit) {
```

Comment:
The type IDs should already be sorted. You could use `java.util.Arrays#binarySearch(int[], int)`, although for such a small size it probably won't make a lot of difference.

Reply:
I figured linear search would be faster and potentially even loop unrolled if the compiler can prove cardinality <= MAX_CARDINALITY. I would have to actually investigate this